### PR TITLE
Add system for safely manipulating JSON databases and apply it to photo albums and photo frames

### PR DESCRIPTION
**Pull request by Mothblocks**
We frequently have issues with data loss in our long storage .json files for various reasons, such as the file being completely blanked out on write etc. 

This introduces a system that tries to safely handle that by saving the known working json file into a backup that will be loaded in the case a write fails.

This system queues updates in order to send through to the next tick. This is an improvement over the existing implementation of photo albums and photo frames (I think all persistence, even) which do not save until the end of a properly rebooted round, but not during a server crash.

Also saves the jsons in pretty prints, which make them easier to read but especially make them easier to diff in a git repository, which MSO wants to setup (and hopefully make public so I can make a dashboard on bus.moth.fans for looking at photo albums and their history, which is something I've wanted to do for a very long time).

## Changelog
:cl:
refactor: Photo albums and photo frames are now more resilient to data loss, especially when a server crashes.
/:cl:

### CODE REVIEW

File: code/datums/json_database.dm [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+/// Removes the data at the given item, and queues a save.
+/// For dictionaries, this can be the key.
+/// For arrays, this can be the value.
+/datum/json_database/proc/remove(item)
+	cached_data -= item
+	queue_save()
+
+/// Inserts the data at the end of what is assumed to be an array, and queues a save.
+/datum/json_database/proc/insert(value)
+	UNTYPED_LIST_ADD(cached_data, value)
```
Comment (MrStonedOne) [inline code/datums/json_database.dm:74]:
```suggestion
	UNTYPED_LIST_REMOVE(cached_data, value)
	queue_save()

/// Inserts the data at the end of what is assumed to be an array, and queues a save.
/datum/json_database/proc/insert(value)
	UNTYPED_LIST_ADD(cached_data, value)
```

the entire reason i ended up making `UNTYPED_LIST_ADD` and friends was when i found out the same issue existed for `-=` and lists and realized how that might not have been known.

File: code/datums/json_database.dm [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	VAR_PRIVATE
+		filepath
+		backup_filepath
+
+		cached_data
+		save_queued = FALSE
+
+		static/existing_json_database = list()
+
+/datum/json_database/New(filepath)
```
Comment (MrStonedOne) [inline code/datums/json_database.dm:15]: needs admin proc call check.

File: code/datums/json_database.dm [hunk 3]
```diff
@@ -0,0 +1,117 @@
+/// Represents a json file being used as a database in the data/ folder.
+/// Changes made here will save back to the associated file, with recovery.
+/// Will defer writes until later if multiple happen in the same tick.
+/// Do not add an extra cache on top of this. This IS your cache.
+/datum/json_database
+	VAR_PRIVATE
+		filepath
+		backup_filepath
```
Comment (MrStonedOne) [inline code/datums/json_database.dm:8]: needs admin varedit protections

**Review by MrsTonedOne — APPROVED** [review]

File: code/modules/photography/photos/frame.dm [hunk 4]
```diff
@@ -113,6 +105,15 @@
...
 		update_appearance()
 
+/// Given a photo (or null), will change the contained picture, and queue a persistent save.
+/obj/structure/sign/picture_frame/proc/set_and_save_framed(obj/item/photo/photo)
+	framed = photo
+
+	if (isnull(persistence_id))
+		return
+
+	SSpersistence.photo_frames_database.set_key(persistence_id, photo.picture?.id)
```
Comment (MrMelbert) [inline code/modules/photography/photos/frame.dm:115]: `photo` can be null but there is no null check for `photo.picture`

Comment (Mothblocks) [reply code/modules/photography/photos/frame.dm:115]:
```suggestion
/// Given a photo (or null), will change the contained picture, and queue a persistent save.
/obj/structure/sign/picture_frame/proc/set_and_save_framed(obj/item/photo/photo)
	framed = photo

	if (isnull(persistence_id))
		return

	SSpersistence.photo_frames_database.set_key(persistence_id, photo?.picture?.id)
```

**Review by Ghommie — APPROVED** [review]

**Pull request merged**